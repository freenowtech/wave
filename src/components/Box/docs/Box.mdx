---
name: Box
menu: Components
route: /components/box
---

import { Playground } from 'docz';
import { Box } from '../Box';
import { StyledSystemLinks } from '../../../docs/StyledSystemLinks'

# Box

This component serves as a wrapper component for most layout related needs. Use Box to set values such as `display`,
`width`, `height`, and more. In practice, this component is used frequently as a wrapper around other components to
achieve Box Model related styling.

<StyledSystemLinks component="Box" supportedProps={ ['space', 'layout', 'position', 'color', 'flexbox', 'grid-layout', 'background'] }/>

## Examples

<Playground>
    Box can be used to create both <Box as="span" bg="#9ED9BB">inline</Box> and
    <Box bg="#9FC5F1">block-level elements,</Box>
    <Box bg="#FF9CFC" width={[1, 1, 1/2]}>elements with fixed or responsive width and height,</Box>
    <Box bg="#FFEB9C" p={4} mt={2}>and more!</Box>
</Playground>
