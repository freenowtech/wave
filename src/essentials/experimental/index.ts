export * from './theme';
export { createGlobalStyle, ColorPalette } from './Colors';
export { darkTheme } from './globalStyles';
export { getSemanticValue, getBareValue, getBareHslValue, getSemanticHslValue } from './cssVariables';
