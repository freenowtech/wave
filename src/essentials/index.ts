export { Elevation } from './Elevation/Elevation';
export { Spaces } from './Spaces/Spaces';
export { Breakpoints, MediaQueries } from './Breakpoints/Breakpoints';
export type { ReadCssColorVariable, ReadCssVariable } from './Colors/types';
export { GlobalStyle as ClassicColors } from './Colors/Colors';
export { GlobalStyle as ModernColors } from './Colors/ModernColors';
