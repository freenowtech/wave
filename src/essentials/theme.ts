import { Colors, SemanticColors } from './Colors/Colors';
import { Spaces } from './Spaces/Spaces';
import { Breakpoints, MediaQueries } from './Breakpoints/Breakpoints';

const theme = {
    breakpoints: Breakpoints,
    colors: Colors,
    // todo: rename semanticColors to colors in the next major release
    semanticColors: SemanticColors,
    fontSizes: ['0.75rem', '0.875rem', '1rem', '1.25rem', '1.5rem', '2rem', '2.5rem', '3rem'],
    fontWeights: {
        normal: 400,
        semibold: 600,
        bold: 700
    },
    fonts: {
        normal: fontStack(['Open Sans', 'sans-serif']),
        monospace: fontStack(['Hack', 'monospace'])
    },
    mediaQueries: MediaQueries,
    space: Spaces,
    radii: ['0rem', '0.125rem', '0.25rem', '0.5rem'],
    iconSizes: {
        small: 12,
        medium: 24,
        large: 28
    },
    shadows: {
        small: '0 0.0625rem 0.25rem 0 rgba(0,0,0,0.1)',
        medium: '0 0 0.5rem 0.1875rem rgba(0,0,0,0.08)',
        large: '0 0 0.75rem 0.25rem rgba(0,0,0,0.12)'
    }
};

/**
 * Custom aliases declaration
 * https://styled-system.com/theme-specification/#scale-aliases
 */

// @ts-ignore
theme.fontSizes.small = theme.fontSizes[1];
// @ts-ignore
theme.fontSizes.medium = theme.fontSizes[2];
// TODO: the alias large should match with scale on theme "fontSizes" but it may require a breaking change
// @ts-ignore
theme.fontSizes.large = '1.125rem';

function fontStack(fonts) {
    return fonts.map(font => (font.includes(' ') ? `"${font}"` : font)).join(', ');
}

export { theme };
