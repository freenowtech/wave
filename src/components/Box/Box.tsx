import styled, { StyledComponentBase } from 'styled-components';
import {
    background,
    BackgroundProps,
    color,
    ColorProps,
    compose,
    flexbox,
    FlexboxProps,
    grid,
    GridProps,
    layout,
    LayoutProps,
    position,
    PositionProps,
    space,
    SpaceProps,
    textAlign,
    TextAlignProps
} from 'styled-system';
import { theme } from '../../essentials/theme';

interface BoxProps
    extends SpaceProps,
        LayoutProps,
        PositionProps,
        ColorProps,
        FlexboxProps,
        GridProps,
        BackgroundProps,
        TextAlignProps {}

const Box: StyledComponentBase<'div', never, BoxProps, 'theme'> = styled.div.attrs({ theme })<BoxProps>`
    ${compose(space, layout, position, color, flexbox, grid, background, textAlign)}
`;

export { Box, BoxProps };
