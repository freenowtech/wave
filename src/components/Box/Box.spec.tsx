import { render } from '@testing-library/react';
import * as React from 'react';
import { Colors } from '../../essentials';
import { Box, BoxProps } from './Box';

describe('Box', () => {
    it('renders without any props', () => {
        expect(render(<Box />).container.firstChild).toMatchSnapshot();
    });

    it('respects the "as" prop', () => {
        expect(render(<Box as="span" />)).toMatchHtmlTag('span');
    });

    it('renders padding and margin', () => {
        expect(render(<Box m={2} p={1} />).container.firstChild).toMatchSnapshot();
    });

    it('renders layout', () => {
        expect(render(<Box width="50%" overflowY="auto" />).container.firstChild).toMatchSnapshot();
    });

    it('renders color', () => {
        expect(render(<Box backgroundColor={Colors.POSITIVE_GREEN_900} />).container.firstChild).toMatchSnapshot();
    });

    it('renders flexbox', () => {
        expect(render(<Box alignItems="center" flexDirection="column" />).container.firstChild).toMatchSnapshot();
    });

    it('renders grid layout', () => {
        expect(
            render(<Box gridColumnGap={2} gridTemplateRows="1fr auto 10%" />).container.firstChild
        ).toMatchSnapshot();
    });

    it('accepts props spreading', () => {
        const renderBox = () => {
            const props = { backgroundColor: Colors.POSITIVE_GREEN_900 };
            return <Box {...props} />;
        };
        expect(renderBox).not.toThrow();
    });
});
