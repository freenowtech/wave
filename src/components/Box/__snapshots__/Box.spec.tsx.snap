// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Box renders color 1`] = `
.c0 {
  background-color: #069D4F;
}

<div
  class="c0"
/>
`;

exports[`Box renders flexbox 1`] = `
.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

<div
  class="c0"
/>
`;

exports[`Box renders grid layout 1`] = `
.c0 {
  grid-column-gap: 1rem;
  grid-template-rows: 1fr auto 10%;
}

<div
  class="c0"
/>
`;

exports[`Box renders layout 1`] = `
.c0 {
  width: 50%;
  overflow-y: auto;
}

<div
  class="c0"
  width="50%"
/>
`;

exports[`Box renders padding and margin 1`] = `
.c0 {
  margin: 1rem;
  padding: 0.5rem;
}

<div
  class="c0"
/>
`;

exports[`Box renders without any props 1`] = `
<div
  class=""
/>
`;
